zdm/MCMC.py

```diff
@@ -0,0 +1,141 @@
+    # np.save(outfile,posterior_sample)
+
+    return
+
+#==============================================================================
+
+def calc_log_posterior(param_vals, params, surveys, grids):
+
+    t0 = time.time()
+    # Can use likelihoods instead of posteriors because we only use uniform priors which just changes normalisation of posterior 
+    # given every value is in the correct range. If any value is not in the correct range, log posterior is -inf
+    in_priors = True
+    param_dict = {}
+
+    for i, (key,val) in enumerate(params.items()):
+        if param_vals[i] < val['min'] or param_vals[i] > val['max']:
+            in_priors = False
+            break
+
+        param_dict[key] = param_vals[i]
+
+    if in_priors == False:
+        llsum = -np.inf
+    else:
+
+        # minimise_const_only does the grid updating so we don't need to do it explicitly beforehand
+        try:
+            newC, llC = it.minimise_const_only(param_dict, grids, surveys)
+            for g in grids:
+                g.state.FRBdemo.lC = newC
+
+            # calculate all the likelihoods
+            llsum = 0
+            for s, grid in zip(surveys, grids):
+                # if DMhalo != None:
+                #     s.init_DMEG(DMhalo)
+                if 'DMhalo' in param_dict:
+                    s.init_DMEG(param_dict['DMhalo'])
+
+                llsum += get_likelihood(grid,s)
+
+        except ValueError as e:
+            print("ValueError, setting likelihood to -inf: " + str(e))
+            llsum = -np.inf
+
+    if np.isnan(llsum):
+        print("llsum was NaN. Setting to -infinity", param_dict)    
+        llsum = -np.inf
+    
+    print("Posterior calc time: " + str(time.time()-t0) + " seconds", flush=True)
+
+    return llsum
+
+#==============================================================================
+
+def mcmc_runner(logpf, outfile, params, surveys, grids, nwalkers=10, nsteps=100, ncpus=1):
+    ndim = len(params)
+    starting_guesses = []
+
+    # Produce starting guesses for each parameter
+    for key,val in params.items():
+        starting_guesses.append(st.uniform(loc=val['min'], scale=val['max']-val['min']).rvs(size=[nwalkers]))
+        print(key + " priors: " + str(val['min']) + "," + str(val['max']))
+    
+    starting_guesses = np.array(starting_guesses).T
+
+    backend = emcee.backends.HDFBackend(outfile+'.h5')
+    backend.reset(nwalkers, ndim)
+
+    start = time.time()
+    with mp.Pool(ncpus) as pool:
+        sampler = emcee.EnsembleSampler(nwalkers, ndim, logpf, args=[params, surveys, grids], backend=backend, pool=pool)
+        sampler.run_mcmc(starting_guesses, nsteps, progress=True)
+    end = time.time()
+    print("Total time taken: " + str(end - start))
+    
+    posterior_sample = sampler.get_chain()
+
+    return posterior_sample
+
+#==============================================================================
+
+def get_likelihood(grid, s):
+
```

Comment:
Great function! This probably appears in several places, right? I think this function should appear in iteration.py, since it's got more general functionality than MCMC. But happy to keep it here for the moment